core/src/main/java/org/jruby/util/RubyStringBuilder.java

```diff
@@ -0,0 +1,248 @@
+                    prev = ++p;
+                }
+                continue;
+            }
+            n = MBCLEN_CHARFOUND_LEN(n);
+            final int c = enc.mbcToCode(bytes, p, end); int cc = 0;
+
+            p += n;
+            if ((asciiCompat || isUnicode) &&
+                    (c == '"' || c == '\\' ||
+                            (c == '#' &&
+                                    p < end &&
+                                    MBCLEN_CHARFOUND_P(StringSupport.preciseLength(enc, bytes, p, end)) &&
+                                    ((cc = codePoint(runtime, enc, bytes, p, end)) == '$' ||
+                                            cc == '@' || cc == '{')
+                            )
+                    )) {
+                if (p - n > prev) result.cat(bytes, prev, p - n - prev);
+                result.cat('\\');
+                if (asciiCompat || enc == resultEnc) {
+                    prev = p - n;
+                    continue;
+                }
+            }
+
+            switch (c) {
+                case '\n': cc = 'n'; break;
+                case '\r': cc = 'r'; break;
+                case '\t': cc = 't'; break;
+                case '\f': cc = 'f'; break;
+                case '\013': cc = 'v'; break;
+                case '\010': cc = 'b'; break;
+                case '\007': cc = 'a'; break;
+                case 033: cc = 'e'; break;
+                default: cc = 0; break;
+            }
+
+            if (cc != 0) {
+                if (p - n > prev) result.cat(bytes, prev, p - n - prev);
+                result.cat('\\');
+                result.cat(cc);
+                prev = p;
+                continue;
+            }
+
+            // FIXME: Can't use Encoding.isAscii because it does not treat int as unsigned 32-bit
+            if ((enc == resultEnc && enc.isPrint(c)) || (asciiCompat && (c < 128 && c > 0) && enc.isPrint(c))) {
+                continue;
+            } else {
+
+                if (p - n > prev) result.cat(bytes, prev, p - n - prev);
+                String format =  StringSupport.escapedCharFormat(c, isUnicode);
+                if (!format.equals("%c")) needsQuotes = true;
+                Sprintf.sprintf(runtime, result.getByteList() ,format, (c & 0xFFFFFFFFL));
+                prev = p;
+                continue;
+            }
+        }
+
+        if (p > prev) result.cat(bytes, prev, p - prev);
+
+        // FIXME: bytelist_love: This is weird and should be done differently
+        if (needsQuotes) {
+            RubyString newResult = runtime.newString("\"");
+            newResult.cat(result);
+            newResult.cat('"');
+            result = newResult;
+        }
+        return result;
+    }
+    
+    public static String str(Ruby runtime, IRubyObject value, String message) {
+        RubyString buf = (RubyString) value.asString().dup();
+
+        buf.cat19(runtime.newString(message));
```

Comment:
do we still need to be using 19 legacy named methods? maybe `cat` + `cat19` are worth some 'unification' ...

Reply:
Unfortunately, they are very different and both should always exist.  I would agree with you that we should refactor with better names as cat19 means nothing but I suspect cat19 is still used by multiple native extensions which use m17n strings in any way (although we can definitely come up with a new name and deprecate).  Main difference is one could care less about what it is cat'ing and the other realizes it needs encoding negotiation.

Reply:
ah yeah I recall them being impossible to unify now ... thanks